### ISSUE TITLE
Container Image size is too big for lambda

### ISSUE THREAD

**Issue by Rionel-Dmello** [aws-samples/aws-genai-llm-chatbot#38]
Hi, When deploying this application with all the RAG stacks enabled, I get this error from cloudformation during the deployment 
```
"Lambda function AwsGenaiLllmChatbotStack-OpenSearchVectorSearchInd-RvPI85ZArJDJ reached terminal FAILED state due to InvalidImage(SizeLimitExceeded: Uncompressed container imag
e size exceeds 10 GiB limit) and failed to stabilize"
```
I believe that we exceed the maximum size of lambda containers.  This happens for the Aurora and Opensearch RAG sources.

**Comment by JahnKhan** [comment]
Hi, i enabled only pgvector RAG source and the deployment is stuck during the publish process. Have you experienced something similar? It just takes hours and then runs into a timeout. It does retrying but without success..

**Comment by gregsomm** [comment]
I hit the same error when attempting to use OpenSearch:

```
34/53 Currently in progress: AwsGenaiLllmChatbotStack, OpenSearchVectorSearchApiHandler6B246D88, OpenSearchVectorSearchCreateIndex7EED590A, OpenSearchVectorSearchIndexDocument36B9A569
AwsGenaiLllmChatbotStack | 34/53 | 2:52:03 PM | CREATE_FAILED        | AWS::Lambda::Function                           | OpenSearchVectorSearch/IndexDocument (OpenSearchVectorSearchIndexDocument36B9A569) Resource handler returned message: "Lambda function AwsGenaiLllmChatbotStack-OpenSearchVectorSearchInd-7ZOZdnMIEY1f reached terminal FAILED state due to InvalidImage(SizeLimitExceeded: Uncompressed container image size exceeds 10 GiB limit) and failed to stabilize" (RequestToken: f8e7a35a-da2e-bc26-b3d6-af6c6f55add2, HandlerErrorCode: NotStabilized)
AwsGenaiLllmChatbotStack | 34/53 | 2:52:04 PM | CREATE_FAILED        | AWS::Lambda::Function                           | OpenSearchVectorSearch/ApiHandler (OpenSearchVectorSearchApiHandler6B246D88) Resource creation cancelled
AwsGenaiLllmChatbotStack | 34/53 | 2:52:04 PM | CREATE_FAILED        | AWS::Lambda::Function                           | OpenSearchVectorSearch/CreateIndex (OpenSearchVectorSearchCreateIndex7EED590A) Resource creation cancelled
```
Any clues about what's happening?  I assume one of the packages has grown recently, and that's causing the build to exceed 10 GB...

**Comment by chloe-kwak** [comment]
I also hit the same error when creating PGVector  during deplyment due to the container size (over 10GiB). 
`1:57:39 PM | CREATE_FAILED        | AWS::Lambda::Function                           | AuroraPgVector/Doc...g/DocumentIndexing
Resource handler returned message: "Lambda function AwsGenaiLllmChatbotStack-AuroraPgVectorDocumentInd-GSLQpWzr9KTr reached terminal FAILED state due to InvalidImage(SizeL
imitExceeded: Uncompressed container image size exceeds 10 GiB limit) and failed to stabilize" (RequestToken: aba5d2f6-c030-ba0a-d5e9-189946108dbc, HandlerErrorCode: NotSt
abilized)
1:57:43 PM | ROLLBACK_IN_PROGRESS | AWS::CloudFormation::Stack                      | AwsGenaiLllmChatbotStack
The following resource(s) failed to create: [OpenSearchVectorSearchOpenSearchCollectionE26DF627, AuroraPgVectorDocumentIndexingEDBFE8E5, AuroraPgVectorApiHandler56DED5F2,
AuroraPgVectorSemanticSearchApi8E1D9826]. Rollback requested by user.
1:57:43 PM | ROLLBACK_IN_PROGRESS | AWS::CloudFormation::Stack                      | AwsGenaiLllmChatbotStack
The following resource(s) failed to create: [OpenSearchVectorSearchOpenSearchCollectionE26DF627, AuroraPgVectorDocumentIndexingEDBFE8E5, AuroraPgVectorApiHandler56DED5F2,
AuroraPgVectorSemanticSearchApi8E1D9826]. Rollback requested by user.`

**Comment by QuinnGT** [comment]
Same error. 
```
Resource handler returned message: "Lambda function AwsGenaiLllmChatbotStack-OpenSearchVectorSearchInd-jlhdiMLnzFoE reached terminal FAILED state due to Inv
alidImage(SizeLimitExceeded: Uncompressed container image size exceeds 10 GiB limit) and failed to stabilize" (RequestToken: c9434bc1-73a0-2b42-b9a5-9ab1993
71783, HandlerErrorCode: NotStabilized)
```

**Comment by MirandaDora** [comment]
same issue

**Comment by gregsomm** [comment]
FYI, a colleague of mine found that locking to previous versions is a workaround:

```
aws_lambda_powertools
aws_xray_sdk
boto3-1.28.21-py3-none-any.whl
botocore-1.31.21-py3-none-any.whl
langchain==0.0.288
opensearch-py
requests_aws4auth
unstructured[all-docs]==0.10.14

```

However, I will say that this is NOT a good solution, since you never get later versions of the libraries in question.  But if you need to get something working right away, this might help.

**Comment by Rionel-Dmello** [comment]
Hi @bigadsoleiman I tried pulling the latest commit and building it. It still fails when deploying the pgvector stack with the error `Lambda function AwsGenaiLllmChatbotStack-AuroraPgVectorDocumentInd-m9rbn5M19Wdn reached terminal FAIL
ED state due to InvalidImage(SizeLimitExceeded: Uncompressed container image size exceeds 10 GiB limit) and failed to stabilize" (Request
Token: 7a72dca0-4d75-d02e-a14e-f3163f4d4478, HandlerErrorCode: NotStabilized)`

**Comment by Rionel-Dmello** [comment]
Okay so with the combination of @bigadsoleiman fix and also changing langchain==0.0.288 everywhere. Also changed the Dockerfile for Opensearch to
`# Use base image
FROM public.ecr.aws/lambda/python:3.11
# Install system dependencies and remove cache
RUN yum update -y && \
    yum install -y git wget gcc rustc rustup cargo libxml2-devel libxslt-devel libmagic-dev poppler-utils tesseract-ocr libreoffice pandoc && \
    yum clean all && \
    rm -rf /var/cache/yum
# Copy only necessary files
COPY dependencies/* ./
COPY requirements.txt ./
COPY index.py ./
# Install Python packages and avoid caching
# Install Python packages and avoid caching
RUN pip install --upgrade pip && \
    pip install --no-cache-dir unstructured[feature1,feature2] && \
    pip install --no-cache-dir -r requirements.txt --upgrade
# Download NLTK resources
RUN python -c "import nltk;nltk.download('punkt', download_dir='/home/sbx_user1051/nltk_data')" && \
    python -c "import nltk;nltk.download('averaged_perceptron_tagger', download_dir='/home/sbx_user1051/nltk_data')"
# Set CMD
CMD ["index.lambda_handler"]`

**Comment by QuinnGT** [comment]
> Okay so with the combination of @bigadsoleiman fix and also changing langchain==0.0.288 everywhere. Also changed the Dockerfile for Opensearch to
> 
> ```
> FROM public.ecr.aws/lambda/python:3.11
> RUN yum update -y && \
>     yum install -y git wget gcc rustc rustup cargo libxml2-devel libxslt-devel libmagic-dev poppler-utils tesseract-ocr libreoffice pandoc && \
>     yum clean all && \
>     rm -rf /var/cache/yum
> COPY dependencies/* ./
> COPY requirements.txt ./
> COPY index.py ./
> RUN pip install --upgrade pip && \
>     pip install --no-cache-dir unstructured[feature1,feature2] && \
>     pip install --no-cache-dir -r requirements.txt --upgrade
> RUN python -c "import nltk;nltk.download('punkt', download_dir='/home/sbx_user1051/nltk_data')" && \
>     python -c "import nltk;nltk.download('averaged_perceptron_tagger', download_dir='/home/sbx_user1051/nltk_data')"
> CMD ["index.lambda_handler"]
> ```

@Rionel-Dmello did this resolve it the size issue or were you just clarifying this is what you used that generated the error?

**Comment by Rionel-Dmello** [comment]
@QuinnGT Thanks! Edited my original post, but this helped me deploy both the Kendra and the Opensearch stacks. PGVector still throws an error.

**Comment by gonzobrandon** [comment]
Same. for pgVector & version freeze (and git pull master to latest with freeze) container is > 10GB, rollback:

```
Lambda function AwsGenaiLllmChatbotStack-AuroraPgVectorDocumentInd-ybhM3EDcYGDg reached terminal FAILED state due to InvalidImage(SizeLimitExceeded
: Uncompressed container image size exceeds 10 GiB limit) 
```

**Comment by jawhnycooke** [comment]
Same issue as the above reports
Lambda function AwsGenaiLllmChatbotStack-AuroraPgVectorDocumentInd-27OdlB7gdQ0g reached terminal FAILED state due to InvalidImage(SizeLimitExceeded: Uncompressed container image size exceeds 10 GiB limit) and failed to stabilize

**Comment by awsglen** [comment]
Has anyone gotten past this issue?   I froze langchain at 0.0.288 and unstructured at 0.10.14 everywhere but still get the AwsGenaiLllmChatbotStack-AuroraPgVectorDocumentInd-27OdlB7gdQ0g reached terminal FAILED state due to InvalidImage(SizeLimitExceeded: Uncompressed container image size exceeds 10 GiB limit) error. The above Dockerfile's error out for me.  What else can be checked to get by this error?

**Comment by schadem** [comment]
I can confirm that changing both (like @Rionel-Dmello mentioned), the requrements.txt and the Dockerfile, fixes the issue for me as well

> Okay so with the combination of @bigadsoleiman fix and also changing langchain==0.0.288 everywhere, I was able to deploy the Kendra and Opensearch stacks successfully. PGVector still fails with the size > 10 GB. Also changed the Dockerfile for Opensearch to the file below with @koushal2018 suggestion
> 
> ```
> FROM public.ecr.aws/lambda/python:3.11
> RUN yum update -y && \
>     yum install -y git wget gcc rustc rustup cargo libxml2-devel libxslt-devel libmagic-dev poppler-utils tesseract-ocr libreoffice pandoc && \
>     yum clean all && \
>     rm -rf /var/cache/yum
> COPY dependencies/* ./
> COPY requirements.txt ./
> COPY index.py ./
> RUN pip install --upgrade pip && \
>     pip install --no-cache-dir unstructured[feature1,feature2] && \
>     pip install --no-cache-dir -r requirements.txt --upgrade
> RUN python -c "import nltk;nltk.download('punkt', download_dir='/home/sbx_user1051/nltk_data')" && \
>     python -c "import nltk;nltk.download('averaged_perceptron_tagger', download_dir='/home/sbx_user1051/nltk_data')"
> CMD ["index.lambda_handler"]
> ```

**Comment by bigadsoleiman** [comment]
We released a new version `v.3.0.0` in which document ingestions is handled with AWS Batch instead of AWS Lambda to accomodate growing dependencies and possible timeout issues.

V2 won't be supported so we suggest to move to V3 for a more long term solutions.
